_posts/2022-04-09-how-eex-turns-your-template-into-html.md

```diff
@@ -0,0 +1,288 @@
+
+The tokens also include some row/column metadata, which is helpful later own for attributing error messages to the origin point in the source code.
+
+Now that we have the tokens, we need to traverse over them and pass them into the engine.
+
+## The Engine
+
+The engine is where the magic happens! It's responsible for building the tokens into usable Elixir AST, which can then be injected into a function.
+
+An `EEx.Engine` has the following callbacks that it must implement, and that your compiler must call appropriately. I'll copy/paste some of the callback documentation here, copyright belonging to the elixir-lang/elixir project.
+
+```elixir
+@doc """
+Called at the beginning of every template.
+It must return the initial state.
+"""
+@callback init(opts :: keyword) :: state
+
+@doc """
+Called at the end of every template.
+It must return Elixir's quoted expressions for the template.
+"""
+@callback handle_body(state) :: Macro.t()
+
+@doc """
+Called for the text/static parts of a template.
+It must return the updated state.
+"""
+@callback handle_text(state, [line: pos_integer, column: pos_integer], text :: String.t()) ::
+            state
+
+@doc """
+Called for the dynamic/code parts of a template.
+The marker is what follows exactly after `<%`. For example,
+`<% foo %>` has an empty marker, but `<%= foo %>` has `"="`
+as marker. The allowed markers so far are:
+  * `""`
+  * `"="`
+  * `"/"`
+  * `"|"`
+Markers `"/"` and `"|"` are only for use in custom EEx engines
+and are not implemented by default. Using them without an
+appropriate implementation raises `EEx.SyntaxError`.
+It must return the updated state.
+"""
+@callback handle_expr(state, marker :: String.t(), expr :: Macro.t()) :: state
+
+@doc """
+Invoked at the beginning of every nesting.
+It must return a new state that is used only inside the nesting.
+Once the nesting terminates, the current `state` is resumed.
+"""
+@callback handle_begin(state) :: state
+
+@doc """
+Invokes at the end of a nesting.
+It must return Elixir's quoted expressions for the nesting.
+"""
+@callback handle_end(state) :: Macro.t()
+```
+
+The compiler will start by calling the `c:init` callback to initialize the engine's state. It will pass this state to the various other callbacks as required. The state is an implementation detail to the Engine but will usually collect the static and dynamic pieces of the template as it traverses the tokens.
+
+Whenever the compiler encounters a "text" token, it will pass the state, the metadata, and the text to he `c:handle_text/3` callback.
+
+Whenever the compiler encounters an `expression` token, it will pass the state, the "marker", and the expression AST to the `c:handle_expr/3` callback.
+
+If the encountered expression is multi-line, it will call the `c:handle_begin/3` callback. This creates a new "nesting", and will will return a brand new state and should collect the output for the "nested" tokens.
```

Comment:
```suggestion
If the encountered expression is multi-line, it will call the `c:handle_begin/3` callback. This creates a new "nesting" and will return a brand new state. It should also collect the output for the "nested" tokens.
```